### DISCUSSION TITLE
Possible to modify a sensor value before it's sent?

### DISCUSSION THREAD

**Discussion by papakpmartin** [arendst/Tasmota#20612] [Q&A] [answered]
I have a custom ESP32-S3 board with a SHTC3 temp/humidity sensor on it. All works nicely with Tasmota, but the temp sensor reads too high because the board runs slightly warm.

Is there a way to "do some math" on the sensor value before it gets sent out?

Thanks! :)

**Comment by sfromis** [comment, marked as answer]
Only thing is the `TempOffset` command.
With ESP32 and Berry, you could also code a driver to an additional sensor object in the JSON payloads.

More generally, the better solution would be to fix your broken board design. A temperature sensor needs both physical distance, and avoidance of heat flows via materials or air. You cannot reliably fix such heat transfer via an offset, or other calculations.

**Reply by papakpmartin** [reply]
`TempOffset`! Awesome... Tasmota always impress me. :)

Yes, the board isn't right, for sure, but for my rudimentary purposes it seem as though `TempOffset` is going to keep me close enough.

Thanks so much, @sfromis!

**Comment by lalo-uy** [comment]
If you need a more complex calc than Offset, you can capture the value with
a rule , adjust and send it on a nuw topic.

**Reply by papakpmartin** [reply]
Thanks @lalo-uy! Yeah, I'm quite new to Berry (did even really know it was in there), so it's now on "my list", but the offset is working nicely so far to get me closer.

**Reply by sfromis** [reply]
To clarify, using a rule and Berry are not the same. Berry is a full programming language, broadly similar to a simplified Python, and included with Tasmota on ESP32. The older rules feature is not like a programming language, but instead a scheme where you can react to events and then issue commands. Can be tricky for more than simple automations, but is also available on the lower-capacity ESP8266.

When I specifically talked about the option of using Berry, that was for something where you can not do quite the same in rules. With Berry, you can code a sensor driver, on par with the drivers natively part of Tasmota core. This can be processed in the same way as regular sensors, including sensor-based automation, and you could have a driver picking up the value from a "real" sensor, and then show up as another sensor (with a different name) in the same payload, albeit not represening the exact same point in time.

**Reply by papakpmartin** [reply]
Thanks @sfromis... I definitely was conflating rules and Berry because I've used neither. But what I envision that I'd _like_ to do is the Berry sensor driver idea. That makes good, intuitive sense to me. I just haven't learned how yet. :)
